internal/v2/security.go

```diff
@@ -0,0 +1,119 @@
+package v2
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+)
+
+// SecurityProviderExoscaleV2 represents an Exoscale API V2 security provider.
+type SecurityProviderExoscaleV2 struct {
+	// ReqExpire represents the request expiration duration.
+	ReqExpire time.Duration
+
+	apiKey    string
+	apiSecret string
+}
+
+// NewSecurityProviderExoscaleV2 returns a new Exoscale API V2 security provider to sign API requests using the
+// specified API key/secret.
+func NewSecurityProviderExoscaleV2(apiKey, apiSecret string) (*SecurityProviderExoscaleV2, error) {
+	var provider = SecurityProviderExoscaleV2{
+		ReqExpire: 10 * time.Minute,
+	}
+
+	if provider.apiKey = apiKey; provider.apiKey == "" {
+		return nil, errors.New("missing API key")
+	}
+
+	if provider.apiSecret = apiSecret; provider.apiSecret == "" {
+		return nil, errors.New("missing API secret")
+	}
+
+	return &provider, nil
+}
+
+// Intercept is an HTTP middleware that intercepts and signs client requests before sending them to the API
+// endpoint.
+func (s *SecurityProviderExoscaleV2) Intercept(_ context.Context, req *http.Request) error {
+	return s.signRequest(req, time.Now().UTC().Add(s.ReqExpire))
+}
+
+func (s *SecurityProviderExoscaleV2) signRequest(req *http.Request, expiration time.Time) error {
+	var buf bytes.Buffer
+
+	authHeader := fmt.Sprintf("EXO2-HMAC-SHA256 credential=%s", s.apiKey)
+
+	// Request method/URL path
+	if _, err := buf.WriteString(fmt.Sprintf("%s %s\n", req.Method, req.URL.Path)); err != nil {
+		return err
+	}
+
+	// Request body
```

Comment:
I'd move this into a dedicated function. `slurpRequestBody` maybe ?

Reply:
The change to concatenating signature parts instead of manually writing to a buffer reduced the global volume of code to a point where I find it overkill to extract this to a dedicated function, especially since it's not reused anywhere...